File: rt/caml/jSOO.ml
(* JavaScript Objects Operations *)

(** This operator tries to simulate the dot (.) of JavaScript. Its
    left operand being the subject object and its right one the
    message to send. *)
let (>>>) x f = f x

(* types *)

(** The abstract type obj represents a JS object *)
type obj ;;

(** The value type is used as an intermediate container to pass values
    between OCaml and JS *)
type value =
  | Obj of obj       (** identity *)
  | Num of float     (** javascript number / OCaml float *)
  | String of string (** strings *)
  | Block of Obj.t   (** identity, but checks if OCaml value *)
  | Nil              (** JS null *)
;;

(* object creators *)

(** creates An empty object *)
external new_obj : obj -> obj = "jsoo_new" ;;
(** Evaluates a JS code *)
external eval : string -> obj = "jsoo_eval" ;;
(** Transforms an OCaml value into a JS object *)
external inject : value -> obj = "jsoo_inject" ;;
(** Extracts an OCaml value from a JS object *)
external extract : obj -> value = "jsoo_extract" ;;
(** null JS value *)
let null = inject Nil ;;
(** Obtain a JS string from a string. The result is a copy so any
    modification to the original string does not affect the JS
    string. *)
let string s = inject (String s) ;;
(** Obtain a JS number from a float value *)
let float f = inject (Num f) ;;
(** Obtain a JS number from an int value *)
let int i = inject (Num (float_of_int i)) ;;

(* object extractors *)

(** extracts a string from a JS object, raises (Failure "as_string")
    in case of error *)
let as_string x = match extract x with String s -> s | _ -> failwith "as_string" ;;
(** extracts an object from a JS object, raises (Failure "as_obj") in
    case of error *)
let as_obj x = match extract x with Obj o -> o | _ -> failwith "as_obj" ;;
(** extracts a, int from a JS object, raises (Failure "as_int") in
    case of error *)
let as_int x = match extract x with Num f -> int_of_float f | _ -> failwith "as_int" ;;
(** extracts a floatfrom a JS object, raises (Failure "as_float") in
    case of error *)
let as_float x = match extract x with Num f -> f | _ -> failwith "as_float" ;;
(** extracts a block from a JS object, raises (Failure "as_block") in
    case of error *)
let as_block x = match extract x with Block b -> b | _ -> failwith "as_block" ;;

(* field accessors *)

(** Access a property of a JS object, as a JS object. Parameters are
    reversed to be used with the (>>>) combinator defined above. For
    instance (o >>> get "f") is equivalent to o.f (or o["f"]) in JS *)
external get : string -> obj -> obj = "jsoo_get" ;;
(** Modify a property : (o >>> set "f" v) is equivalent to o.f = v in
    JS *)
external set : string -> obj -> obj -> unit = "jsoo_set" ;;
(** removes a property from an object *)
let unset field obj = set field obj null ;;

(* call JS functions *)

(** Generic call mechanism, takes the function object, the subject
    (bound to 'this' in the body and an array of JS objects to use as
    arguments *)
external call : obj -> obj array -> obj -> obj = "jsoo_call" ;;
(** Calls a method from an object *)
let call_method field args dest =
  let meth = get field dest in
    call dest args meth ;;
(** Calls a function object with a null subject *)
let call_function args f =
  call null args f ;;

(* build JS event handlers from caml closures *)

(** internal function *)
external wrap_event : (unit -> unit) -> obj = "jsoo_wrap_event" ;;
(** internal function *)
external get_event_arg : unit -> obj = "jsoo_get_event_args" ;;
(** Wraps an OCaml functional value into a JS object useable as an
event handler *)
let wrap_event f =
  wrap_event
    (fun () ->
       try
	 f (get_event_arg ()) ;
	 Thread.exit ();
       with e ->
	 Thread.thread_uncaught_exception e;
	 Thread.exit ())
;;


(* meta *)

external current_vm : unit -> obj = "current_vm"

File: rt/js/jsoo.js
// Caml name:  new_obj
// Caml type:  unit -> obj
function jsoo_new (o) {
    return [];
}
    
// Caml name:  eval
// Caml type:  string -> obj
function jsoo_eval (s) {
    try {
	var code = string_from_value (s) ;
	return eval (code);
    } catch (e) {
	caml_catch(e);
	this.failwith("jsoo_call: " + e.message);
    }
}

// Caml name:  get
// Caml type:  string -> obj -> obj
function jsoo_get (f, o) {
    return o[string_from_value (f)];
}

// Caml name:  set
// Caml type:  string -> obj -> obj -> unit
 function jsoo_set (f, v, o) {
     o[string_from_value (f)] = v;
     return UNIT;
}

// Caml name:  extract
// Caml type:  obj -> value
function jsoo_extract (o) {
    //   | Obj of obj        0
    //   | Num of float      1
    //   | String of string  2
    //   | Block of Obj.t    3
    //   | Nil
    if (o == null)
	return 0;
    if (o instanceof Block) {
	var b = mk_block (1, 3);
	b.set (0, o);
	return b;
    }
    if (typeof o == 'string') {
	var b = mk_block (1, 2);
	b.set (0, value_from_string (o));
	return b;
    }
    if (typeof o == 'number') {
	var b = mk_block (1, 1);
	b.set (0, box_float (o));
	return b;
    }
    var b = mk_block (1, 0);
    b.set (0, o);
    return b;
}

// Caml name:  extract_bool
// Caml type:  obj -> bool
function jsoo_extract_bool (o) {
    if (o)
	return 1;
    return 0;
}

// Caml name:  inject
// Caml type:  value -> obj
function jsoo_inject (o) {
    if (!is_block(o))
	return null;
    if (o.tag == 2)
	return string_from_value (o.get(0));
    if (o.tag == 1)
	return unbox_float (o.get(0));
    return o.get (0);
}

// Caml name:  call
// Caml type:  obj -> obj array -> obj -> obj
function jsoo_call (d, args, o) {
    try {
	return o.apply (d, args.content) ;
    } catch (e) {
	caml_catch(e);
	this.failwith("jsoo_call: " + e.message);
    }
}

// Caml name:  wrap_event
// Caml type:  (unit -> unit) > obj
function jsoo_wrap_event (clos, res) {
    var vm = this;
    return function (evt) {
	var pid = vm.thread_new (clos);
	var p = vm.ctx;
	do {
	    if (p.pid == pid) {
		p.event_args = evt;
		break;
	    }
	    p = p.n_ctx;
	} while (p != vm.ctx);
	vm.run ();
	return null;
    }
}

// Caml name:  get_event_args
// Caml type:  unit -> obj array
function jsoo_get_event_args (unit) {
    return this.ctx.event_args;
}

// Type: unit -> obj
function current_vm () {
    return this;
}
